server/matchmaker/matchmaker_queue.py

```diff
@@ -4,15 +4,33 @@
 from datetime import datetime, timezone
 from statistics import mean
 from typing import Deque, Dict
+import time
+import functools
 
 import server
+import server.metrics as metrics
 
 from ..decorators import with_logger
 from .algorithm import make_matches
 from .pop_timer import PopTimer
 from .search import Match, Search
 
 
+def timed_async_search(func):
```

Comment:
Why does this need to be a decorator?

We also might want to know whether the search ended because it found a match or because it was cancelled. A nice way I can think of for doing this would be with a context manager, especially since context managers can see if an exception was thrown. https://docs.python.org/3/reference/datamodel.html#object.__exit__

Then the search code might look something like:
```python
try:
    with metrics.matchmaker_search_timer():
        # Wait for search to complete
except CancelledError:
    pass
finally:
    # Do the cleanup
```

and the context manager would be able to detect if `CancelledError` was thrown or not and label the observation appropriately.

Reply:
Yes, that is indeed nicer.